app/src/main/kotlin/com/mapbox/vision/teaser/replayer/SessionsAdapter.kt

```diff
@@ -1,42 +1,182 @@
+                }
+                if (item.itemType == CAMERA) {
+                    clickCameraListener.invoke()
+                } else {
+                    clickSessionListener.invoke(item.itemName)
+                }
+            }
+        }
+    }
 
-        init {
-            textView.setOnClickListener {
-                clickListener.invoke(sessions[adapterPosition])
+    private fun handleMultiSelectionClick(holder: SessionViewHolder, item: SourceItem) {
+        if (selectedItems.contains(item.itemName)) {
+            holder.itemView.setBackgroundColor(backgroundColorTransparent)
+            holder.iconChecked.setImageResource(R.drawable.ic_not_checked)
+            selectedItems.remove(item.itemName)
+        } else {
+            holder.itemView.setBackgroundColor(backgroundColorSelection)
+            holder.iconChecked.setImageResource(R.drawable.ic_checked)
+            selectedItems.add(item.itemName)
+        }
+    }
+
+    private fun bindViewHolderLongClick(holder: SessionViewHolder, item: SourceItem) {
+        if (item.itemType != CAMERA) {
+            holder.itemView.setOnLongClickListener {
+                if (isMultiSelection) {
+                   handleMultiSelectionClick(holder, item)
+                } else {
+                    selectedItems.add(item.itemName)
+                    onActivateMultiSelectionListener.invoke()
+                }
+                true
             }
         }
     }
+
+    fun activateMultiSelection() {
+        isMultiSelection = true
+        notifyDataSetChanged()
+    }
+
+    fun resetMultiSelection() {
+        isMultiSelection = false
+        selectedItems.clear()
+        notifyDataSetChanged()
+    }
+
+    fun selectAll() {
+        items.forEach {
+            if (it.itemType == SESSION) {
+                selectedItems.add(it.itemName)
+            }
+        }
+        notifyDataSetChanged()
+    }
+
+    fun updateSessionsList() {
+        items.clear()
+        items.add(SourceItem(CAMERA, cameraString, null))
+        File(basePath).listFiles().forEach {
+            val dateString = dateFormatter.format(it.lastModified())
+            items.add(SourceItem(SESSION, it.name, dateString))
+        }
+    }
+
+    override fun getItemCount() = items.size
+
+    inner class SessionViewHolder(view: View) : RecyclerView.ViewHolder(view) {
+        val textSessionName: TextView = view.findViewById(R.id.session_view_session_name)
+        val textSessionDate: TextView = view.findViewById(R.id.session_view_session_date)
+        val iconChecked: ImageView = view.findViewById(R.id.session_view_checked)
+    }
+
+    companion object {
+        @SuppressLint("ConstantLocale")
+        private val dateFormatter = SimpleDateFormat("MMM dd, yyyy HH:mm", Locale.getDefault())
+    }
+
+    enum class ItemType {
+        CAMERA,
+        SESSION
+    }
+
+    data class SourceItem (val itemType: ItemType, val itemName:String, val fileDate: String?)
```

Comment:
Could we use `sealed class` here instead of having `enum class ItemType`?

Reply:
yeah, I'll change

Reply:
fixed